---
title: "Introduction: SpECtOR (SEquence COverage Roughness)"
author: "Anas A Rana"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{SpECtOR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Background

Recent advances in high-throughput sequencing of cancer genomes has elucidated the landscape of genetic alterations in many tumor types, thus expanding molecular classification of cancer and facilitating routine whole genome sequencing in clinical oncology.

The quality of tumor genome is a function of many variables, such as tissue quality and quantity, tumor content within a sample, depth of sequencing and effectiveness of the data processing pipeline. Tissue quality relates to the age and size of the tissue and a fixation protocol. The fresh or frozen (FF) tumor material generally yields high-quality tumor genomic DNA. However, in most cases archival specimens stored as formalin-fixed, paraffin-embedded (FFPE) tissue blocks. Formalin can fragment the genomic DNA and introduce chemical modifications during the fixation and storage process that, eventually, lead to false-positive mutation and copy-number alteration calls. As FFPE tissue currently comprise the most accessible clinical option, evaluation of which samples yield successful whole genome sequence is critically important for molecular diagnostics and production-level sequencing.

## Aim

To develop an approach and a metric for evaluation of noisiness (waviness) of sequenced genomic data. The metric should allow for comparison of whole genome sequence datasets (e.g. samples preserved with FFPE and FF protocols).

## Method

We designed an approach that combines classical methods routinely applied in signal processing field (wavelet transform and fractals) and ISO standards for surface texture roughness measurement.

# Using the code

## Running

To run the analysis use `spector()`.

### Id file

The important thing before running `spector` on  your data is to ensure you have created a file containing grouping, id, and baseline information. A sample file is provided `bam_id.csv`. The file can be `TAB` or `,` delimited, but make sure to change the `f_delim` variable accordingly. There is no header by default, but that can be adjusted with the binary `f_head`.

The variables contained in the file are as follows:

|                 | Description                                                                         |
| :----           | ----                                                                                |
| bam file path*  | full path to that bam file                                                          |
| bam id*         | unique id for that bam sample                                                       |
| group id        | group id only if samples come in groups                                             |
| baseline        | id for the baseline such as germline if it exists                                   |
| prep identifier | if there is a unique identifier grouping such as sample prep used for colour coding |

### Plotting

There are three types of plots that can be generated with this package. You can choose the ones you want to generate by adjusting the `plot_type` variable, which takes a vector. The three options are `"boxplot", "circle", "dot"`.

The __boxplot__ is a simple side-by-side boxplot of all samples.

The __circle__ option will plot circles, where the size central part of the circle represents mean value of the _roughness metric_ across all regions in the sample, the size of the edge of the circle represents the interquartile range of the _roughness metric_ across all regions in a sample. If there is a baseline provided the colour of the each will represent how far they are from the baseline.

The __dot__ option will plot the _roughness metric_ for all regions grouped by chromosome along the horizontal and different samples along the vertical. This allows for an easy visualisation of noise levels in different chromosomes.
